python/graphstorm/model/edge_decoder.py

```diff
@@ -639,3 +639,146 @@ def out_dims(self):
+            if weight_field in g.edges[etype].data:
+                eid = g.edges(form="eid", etype=etype)
+                weight = g.edges[etype].data[weight_field][eid]
+                weight = weight.flatten()
+                assert len(weight) == len(eid), \
+                    "Edge weight must be a tensor of shape (num_edges,) " \
+                    f"or (num_edges, 1). But get {g.edges[etype].data[weight_field].shape}"
+            else: # weight_field not in g, it is a neg_graph
+                weight = th.ones((g.num_edges(etype),))
+        else:
+            # current etype does not has weight
+            weight = th.ones((g.num_edges(etype),))
+    else: # edge_weight_fields is a str
+        weight_field = edge_weight_fields
+        if weight_field in g.edges[etype].data:
+            eid = g.edges(form="eid", etype=etype)
+            weight = g.edges[etype].data[weight_field][eid]
+            weight = weight.flatten()
+            assert len(weight) == len(eid), \
+                 "Edge weight must be a tensor of shape (num_edges,) " \
+                f"or (num_edges, 1). But get {g.edges[etype].data[weight_field].shape}"
+        else:
+            # current etype does not has weight
+            weight = th.ones((g.num_edges(etype),))
+    return weight
+
+class LinkPredictWeightedDistMultDecoder(LinkPredictDistMultDecoder):
+    """Link prediction decoder with the score function of DistMult
+       with edge weight.
+
+       When computing loss, edge weights are used to adjust the loss
+    """
+    def __init__(self, etypes, h_dim, gamma=40., edge_weight_fields=None):
+        self._edge_weight_fields = edge_weight_fields
+        super(LinkPredictWeightedDistMultDecoder, self).__init__(etypes, h_dim, gamma)
+
+    @property
+    def edge_weight_fields(self):
+        """ edge_weight_fields
+        """
+        return self._edge_weight_fields
+
+    def forward(self, g, h):
+        """Forward function.
+
+        This computes the DistMult score on every edge type.
+        """
+        with g.local_scope():
+            scores=[]
+            weights = []
+
+            for canonical_etype in g.canonical_etypes:
+                if g.num_edges(canonical_etype) == 0:
+                    continue # the block might contain empty edge types
+
+                i = self.etype2rid[canonical_etype]
+                self.trained_rels[i] += 1
+                rel_embedding = self._w_relation(th.tensor(i).to(self._w_relation.weight.device))
+                rel_embedding = rel_embedding.unsqueeze(dim=1)
+                src_type, _, dest_type = canonical_etype
+                u, v = g.edges(etype=canonical_etype)
+                src_emb = h[src_type][u]
+
+                dest_emb = h[dest_type][v]
+                rel_embedding = rel_embedding.repeat(1,dest_emb.shape[0]).T
+                scores_etype = calc_distmult_pos_score(src_emb, dest_emb, rel_embedding)
+
+                if self.training:
```

Comment:
why is this needed here, but you don't need it in `LinkPredictWeightedDotDecoder`?

Reply:
We don't actually. Let me remove this.